File: backend/apps/public_website/schemas.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from ninja import Schema


PublicSnapshotTypeValue = Literal[
    "bootstrap",
    "homepage",
    "catalog",
    "content",
]


class PublicWebsitePayloadSchema(Schema):
    environment: str
    generated_at: str
    data: dict


class PublicWebsiteHealthSchema(Schema):
    status: str
    service: str
    timestamp: datetime
    version: str


class PublicSnapshotGenerateSchema(Schema):
    snapshot_type: PublicSnapshotTypeValue
    environment: str = "production"
    ttl_minutes: int = 30


class PublicSnapshotSchema(Schema):
    id: UUID
    snapshot_type: str
    environment: str
    version: int
    payload: dict
    generated_at: datetime
    expires_at: datetime | None
    is_active: bool
    is_expired: bool
    checksum: str



class PublicSnapshotInvalidateSchema(Schema):
    snapshot_type: PublicSnapshotTypeValue | None = None
    environment: str | None = None


class PublicSnapshotRefreshAllSchema(Schema):
    environment: str = "production"
    ttl_minutes: int = 30


class PublicSnapshotInvalidationResultSchema(Schema):
    invalidated_count: int


class PublicSnapshotRefreshResultSchema(Schema):
    environment: str
    snapshot_count: int
    snapshots: list[PublicSnapshotSchema]

File: apps/dashboard/src/features/public-website-snapshots/schemas.ts
import {
  z,
} from "zod";

import {
  snapshotTypes,
} from "./types";

export const snapshotTypeSchema =
  z.enum(snapshotTypes);

export const publicWebsiteSnapshotSchema =
  z.object({
    id: z.string(),
    snapshot_type: snapshotTypeSchema,
    environment: z.string(),
    version: z.number().int().nonnegative(),
    payload: z.unknown(),
    generated_at: z.string(),
    expires_at: z.string().nullable(),
    is_active: z.boolean(),
    is_expired: z.boolean(),
    checksum: z.string(),
  });

export const snapshotListSchema =
  z.array(publicWebsiteSnapshotSchema);

export const invalidationResultSchema =
  z.object({
    invalidated_count:
      z.number().int().nonnegative(),
  });

export const refreshAllResultSchema =
  z.object({
    environment: z.string(),
    snapshot_count:
      z.number().int().nonnegative(),
    snapshots:
      z.array(publicWebsiteSnapshotSchema),
  });

export const publicPreviewSchema =
  z.record(
    z.string(),
    z.unknown(),
  );
